ocaml/typing/ctype.ml

```diff
@@ -3353,31 +3363,42 @@ let unify1_var env t1 t2 =
 (* Called from unify3 *)
 let unify3_var env jkind1 t1' t2 t2' =
   occur_for Unify !env t1' t2;
-  match occur_univar_for Unify !env t2 with
-  | () -> begin
-      unification_jkind_check !env t2' jkind1;
-      link_type t1' t2
-    end
+  (* There are two possible ways forward here. Either the variable [t1']
+     will succeed in unifying with [t2], in which case we're done; or
+     the unification fails, in which case we want to add a GADT equation.
+     But in that second case, we want to pretend that we were making a
+     GADT equation all along, so we don't want any evidence of having
+     attempted the unification (which might, say, fill in a sort variable).
+     We thus backtrack in the GADT equation case. *)
+  let snap = snapshot () in
+  match
+    occur_univar_for Unify !env t2;
+    unification_jkind_check !env t2' jkind1
+  with
+  | () -> link_type t1' t2
   | exception Unify_trace _ when in_pattern_mode () ->
+      backtrack snap;
       reify env t1';
       reify env t2';
       if can_generate_equations () then begin
-        occur_univar ~inj_only:true !env t2';
+        begin match get_desc t2' with
+        | Tconstr(path,[],_)
+          when is_instantiable !env ~for_jkind_eqn:true path ->
```

Comment:
I think this should be `for_jkind_eqn:false`.  You mention you needed this, but it doesn't seem to be needed by the current testsuite, and I think it is wrong: it could result in adding a gadt equation for a built-in type of one defined in the current module, which we otherwise don't allow.  As is discussed in the long comment on `is_instantiable`, it's possible that restriction could one day be dropped with some additional work elsewhere, but I don't see any reason for this one call site to `add_gadt_equation` to be different.

Reply:
Yes. I forget why I changed that. It _was_ necessary, but maybe is no longer now that mcomp is fixed. Perhaps that's what inspired me to explore mcomp. Not worth the time spelunking through all that, though.